src/logger/record.ts

```diff
@@ -0,0 +1,117 @@
+import type { TeamName } from './@types/logger';
+import { isTeam } from './teamStyles';
+
+/** Nominal data */
+type Properties = Record<string, string>;
+/** Numerical data */
+type Metrics = Record<string, number>;
+
+type Flatten<R extends Record<string, string | number>> = Array<{
+	name: keyof R;
+	value: R[keyof R];
+}>;
+
+/** This turns an object’s key-value pairs into an array objects
+ * with `name` and `value` keys.
+ *
+ * @example
+ * ({
+ *   label: 'libs',
+ *   network: 'wifi',
+ * }) => [
+ *   {
+ *		name: 'label',
+ *		value: 'libs',
+ *   },
+ *		name: 'network',
+ *		value: 'wifi',
+ *   },
+ * ]
+ */
+const flatten = <Value extends string | number>(obj: Record<string, Value>) =>
+	Object.entries(obj).map(([name, value]) => ({
+		name,
+		value,
+	}));
+
+type Payload = {
+	label: string;
+	properties: Flatten<Properties>;
+	metrics: Flatten<Metrics>;
+};
+
+const generateJSONPayload = (
+	label: string,
+	properties: Properties,
+	metrics: Metrics,
+): string => {
+	const payload: Payload = {
+		label,
+		properties: flatten(properties),
+		metrics: flatten(metrics),
+	};
+
+	return JSON.stringify(payload);
+};
+
+/** Helper method to get default logging endpoints */
+const getLoggingEndpoint = (
+	isDev: boolean,
+): `https://logs.${string}guardianapis.com/log` =>
+	isDev
+		? 'https://logs.code.dev-guardianapis.com/log'
+		: 'https://logs.guardianapis.com/log';
+
+type Data = {
+	label: `${TeamName}.${string}`;
+	endpoint: string;
+	metrics?: Metrics;
+	properties?: Properties;
+};
+
+/**
+ * Record log to our Data Lake.
+ * ⚠️ Make sure we have consent for collecting this data.
+ * If in doubt, reach out to [@guardian/transparency-consent](https://github.com/orgs/guardian/teams/transparency-consent)
+ *
+ * Send nominal and numerical data:
+ * - Nominal datum is of type `string`
+ * - Numerical datum is of type `number`
+ *
+ * @param {Data} data The data to send
+ * @param {`${TeamName}.${string}`} data.label Used to identify the data (in BigQuery). Starts with a registered team name.
+ * @param {string} data.endpoint The endpoint to send the data to.
+ * @param {Properties} [data.properties] Nominal data. Defaults to an empty object.
+ * @param {Metrics} [data.metrics] Numerical data. Defaults to an empty object.
+ * @returns {boolean} Whether sending the data has been successfully queued.
+ */
+const recordLog = ({
```

Comment:
Can you weave a requirement for getting `ConsentState` for the CMP here, like the commercial-metrics lib does?

See: https://github.com/guardian/commercial-core/blob/main/src/send-commercial-metrics.ts#L157

That would help in making sure consent is respected.

Reply:
Worth noting that even if you respect consent choices here - because consent can be withdrawn retroactively (i.e. only the most recent consent choice matters), I think you'll need to have a filtered view on this table in the data-lake joining to a users current permissions if you want to properly respect consent.